---
layout: post
title: Understanding 10x Xenium Data Analysis
categories: articles
permalink: articles/10x-xenium
author: Atharva Bhagwat
---

## Overview of Spatial Transcriptomics (ST)

Traditional single-cell RNA sequencing (scRNA-seq) dissociates cells and profiles gene expression at single cell resolution. It is able to capture the heterogeneity within tissues and gives high resolution insight into cellular functions.

While this type of data can be used to analyze variations in gene expression, it lacks the spatial information for cells, which adds a dimension of information. Enter ‘Spatial Transcriptomics’ (ST)! ST has revolutionized our ability to understand biology by capturing the spatial information. Having spatial context allows us to look deeper into cell-cell interactions, cellular niches, and various microenvironments across conditions.

Over the last few years, spatial transcriptomics technologies have advanced rapidly and each come with various advantages. Detailed table can be found [here](https://divingintogeneticsandgenomics.com/post/python-visium/).

## Overview of Xenium

Xenium is a single-cell resolution technology commercialized by 10x genomics in 2023. It uses RNA based transcripts and can target up to 5000 genes. It captures transcripts at subcellular level. Xenium v2 has nuclear and cell membrane staining which are used for cell segmentation, as compared to v1 which only had nuclear staining and detected cells based on cell-boundary expansion <ADD REF HERE>.

A few caveots of xenium data are

- Data generation can be baised depending on tissues selected: Data generation for large number of samples can be expensive. So the general practice is to build a Tissue Microarray (TMA) with 30-40 cores of 1.5mm diameter. Selecting these cores haphazardly can induce data bias as it captures a small section of the condition to be captured. So, building TMA with help of expert pathalogists is a necessary step. Ex: When it comes to studying cancer, the most interesting parts are boundaries between either tumor and immune cells or tumor and stromal cells. These regions are full of potential, allowing us to dive deeper into finding interactions and spatial structures between in such regions.

## Understand 10x output files

The Xenium analyzer has numerous outputs we only need some of them to start with data analysis. Here are those files:

| File Type | File and Description |
|------------------------|------------------------|
| **Cell summary** | `cells.parquet`: Cell summary file in Parquet format. |
| **Cell-feature matrix**| `cell_feature_matrix.h5`: Cell-feature matrix file in HDF5 format. |

**Table 2: Xenium Onboard Analysis Output Files ([ref](https://www.10xgenomics.com/support/software/xenium-onboard-analysis/2.0/tutorials/outputs/xoa-output-at-a-glance)).**

## Getting started with data analysis

I will use the [Xenium Human Pancreatic Dataset](https://www.10xgenomics.com/products/xenium-human-pancreatic-dataset-explorer) for showcase. *Please note that this data was generated using Xenium v1, ie: cells are detected using cell boundary expansion.*

```bash
# Output Files
wget https://cf.10xgenomics.com/samples/xenium/1.3.0/Xenium_Preview_Human_Non_diseased_Lung_With_Add_on_FFPE/Xenium_Preview_Human_Non_diseased_Lung_With_Add_on_FFPE_outs.zip
```

Command to unzip:

```bash
unzip -d data Xenium_Preview_Human_Non_diseased_Lung_With_Add_on_FFPE_outs.zip
```

### Installing necessary libraries

```bash
pip install anndata scanpy squidpy
```

### Loading the data as an AnnData object from raw output files

```python

import os
import pandas as pd
import anndata as ad
import scanpy as sc
import squidpy as sq

import matplotlib.pyplot as plt
plt.rcParams['figure.dpi'] = 250        # increase figure dpi for better quality plots

# load count matrix
adata = sc.read_10x_h5('data/cell_feature_matrix.h5')

# load cells metadata
cells = pd.read_parquet('data/cells.parquet')
cells.index = cells['cell_id']
cells.drop(['cell_id'], inplace=True, axis=1)

adata.obs = cells.copy()

# set spatial info
adata.obsm["spatial"] = adata.obs[["x_centroid", "y_centroid"]].copy().to_numpy()

adata.write_h5ad('objects/xenium_adata.h5ad')
```

### Understanding xenium data structure

```python
AnnData object with n_obs × n_vars = 190965 × 474
    obs: 'x_centroid', 'y_centroid', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area'
    var: 'gene_ids', 'feature_types', 'genome'
    obsm: 'spatial'
```

`adata.obs` stores the metadata. `adata.var` stores information about the genes on the panel. More information on AnnData structure can be found here. <ADD REF HERE>
x_centroid and y_xentroid are the x and y coordinates of cells

Run QC:

```python
print(f'Shape of object before filtering: {adata.shape}')

sc.pp.calculate_qc_metrics(adata, percent_top=None, inplace=True)

sc.pp.filter_cells(adata, min_counts=10)
sc.pp.filter_genes(adata, min_cells=5)

print(f'Shape of object after filtering: {adata.shape}')
```

As part of quality control we drop cells with n_genes_by_counts < 10.

```python
AnnData object with n_obs × n_vars = 190455 × 474
    obs: 'x_centroid', 'y_centroid', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'log1p_total_counts', 'n_counts'
    var: 'gene_ids', 'feature_types', 'genome', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_cells'
    obsm: 'spatial'
```

```python
Shape of object before filtering: (190965, 474)
Shape of object after filtering: (190455, 474)
```

For purpose of this article, let's keep things simple and apply the standard scRNA-seq normalization approach.

```python
# Normalize total
sc.pp.normalize_total(adata, target_sum=1e4)

# Log-normalize
sc.pp.log1p(adata)
```

We can apply the same dimensionality reduction methods used for scRNA seq data.

```python
# Run PCA
sc.pp.pca(adata, n_comps=50)

# Calculate neighbors
sc.pp.neighbors(adata, n_neighbors=10, n_pcs=40)

# Leiden clustering
sc.tl.leiden(adata, resolution=0.8)

# Computing UMAP
sc.tl.umap(adata)
```

# Layers of spatial data

*The* reason for spatial data generation is having the x_centroid and y_centroid for each cell. This allows us to build a neighborhood graph.

```python
sq.gr.spatial_neighbors(adata, radius=50, coord_type='generic') # radius in µm
```

This creates an entries in adata.obsp 'spatial_connectivities' and 'spatial_distances' (by default). <ADD REF HERE>

We can now run neighborhood enrichment to see which cells group together using:

```python
# set seed=0 for reproducibility
sq.gr.nhood_enrichment(adata, cluster_key='leiden', seed=0)
```

Focusing on 'spatial_connectivities', we can fetch neighbors for any cell using:

```python
connections = adata.obsp['spatial_connectivities'].tocsr()

def fetch_neighbors(connections, cell_id):
    row_idx = adata.obs_names.get_loc('cell_name')
    neighbor_idx = connections[row_idx].indices
    return adata.obs_names[neighbor_idx]
```

neighborhood and proximity score
